File: feign-core/src/main/java/name/abhijitsarkar/feign/core/matcher/DefaultHeadersMatcher.java
/*
 * Copyright (c) 2016, the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * License for more details.
 */

package name.abhijitsarkar.feign.core.matcher;

import lombok.extern.slf4j.Slf4j;
import name.abhijitsarkar.feign.Request;
import name.abhijitsarkar.feign.core.model.FeignMapping;
import name.abhijitsarkar.feign.core.model.Headers;
import name.abhijitsarkar.feign.core.model.RequestProperties;

import java.util.Locale;
import java.util.Map;
import java.util.function.BiFunction;

import static com.google.common.base.Strings.isNullOrEmpty;
import static com.google.common.base.Strings.nullToEmpty;
import static org.springframework.util.CollectionUtils.isEmpty;

/**
 * @author Abhijit Sarkar
 */
@Slf4j
public class DefaultHeadersMatcher implements BiFunction<Request, FeignMapping, Boolean> {
    @Override
    public Boolean apply(Request request, FeignMapping feignMapping) {
        RequestProperties requestProperties = feignMapping.getRequest();
        Headers headers = requestProperties.getHeaders();
        Map<String, String> requestHeaders = request.getHeaders();

        Map<String, String> pairs = headers.getPairs();

        Boolean ignoreUnknown = headers.isIgnoreUnknown();

        if (!isEmpty(requestHeaders) && isEmpty(pairs) && !ignoreUnknown) {
            log.info("{} {}",
                    "Request headers are not empty but request property headers are,",
                    "and ignoreUnknown is false. Headers match returned false.");

            return false;
        }

        Boolean ignoreEmpty = headers.isIgnoreEmpty();

        if (isEmpty(requestHeaders)) {
            boolean match = ignoreEmpty || isEmpty(pairs);

            log.info("Request headers are empty. Headers match returned {}.", match);

            return match;
        }

        boolean match = requestHeaders.entrySet().stream()
                .allMatch(e -> {
                    String key = e.getKey();
                    String valueFromRequest = e.getValue();

                    log.info("Comparing request header {}:[{}] with {}:[{}].",
                            key, valueFromRequest, key, pairs.get(key));

                    String valueFromRequestProperty = nullToEmpty(pairs.get(key));
                    Boolean ignoreCase = headers.isIgnoreCase();

                    if (ignoreCase) {
                        valueFromRequest = isNullOrEmpty(valueFromRequest) ?
                                valueFromRequest : valueFromRequest.toLowerCase(Locale.ENGLISH);
                        valueFromRequestProperty = valueFromRequestProperty.toLowerCase(Locale.ENGLISH);
                    }

                    boolean m1 = valueFromRequest == null && valueFromRequestProperty.isEmpty();
                    boolean m2 = valueFromRequest != null && valueFromRequestProperty.isEmpty() && ignoreUnknown;
                    boolean m3 = valueFromRequest != null && valueFromRequest.matches(valueFromRequestProperty);

                    if (m1) {
                        log.info("Request header value is null and headers are empty.");
                    }
                    if (m2) {
                        log.info("Request header value is not null, headers are empty and ignoreUnknown is true.");
                    }
                    if (m3) {
                        log.info("Match found.");
                    }

                    return m1 || m2 || m3;
                });

        log.info("Headers match returned {}.", match);

        return match;
    }
}

File: feign-core/src/main/scala/name/abhijitsarkar/feign/core/matcher/DefaultHeadersMatcher.scala
/*
 * Copyright (c) 2016, the original author or authors.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 3 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * A copy of the GNU General Public License accompanies this software,
 * and is also available at http://www.gnu.org/licenses.
 */

package name.abhijitsarkar.feign.core.matcher

import java.util.function.BiFunction

import name.abhijitsarkar.feign.Request
import name.abhijitsarkar.feign.core.model.FeignMapping
import org.slf4j.LoggerFactory

import scala.collection.immutable.{Iterable, Map => ImmutableMap}

/**
  * @author Abhijit Sarkar
  */
class DefaultHeadersMatcher extends BiFunction[Request, FeignMapping, Boolean] {
  val logger = LoggerFactory.getLogger(classOf[DefaultHeadersMatcher])

  override def apply(request: Request, feignMapping: FeignMapping): Boolean = {
    val requestProperties = feignMapping.request
    val headers = requestProperties.headers
    val pairs: ImmutableMap[String, String] = headers.pairsToScala
    val requestHeaders: ImmutableMap[String, String] = request.headers

    val ignoreUnknown = headers.ignoreUnknownToScala.getOrElse(false)

    val isEmpty = (it: Iterable[_]) => it == null || it.isEmpty

    if (!isEmpty(requestHeaders) && isEmpty(pairs) && !ignoreUnknown) {
      logger.info("Request headers are not empty but request property headers are, " +
        "and ignoreUnknown is false. Headers match returned false.")

      return false
    }

    val ignoreEmpty = headers.ignoreEmptyToScala.getOrElse(false)

    if (isEmpty(requestHeaders)) {
      val m = ignoreEmpty || isEmpty(pairs)
      logger.info(s"Request headers are empty. Headers match returned {m}.")

      return m
    }

    val ignoreCase = headers.ignoreCaseToScala.getOrElse(false)
    val maybeToLower = (s: String) => if (ignoreCase) s.toLowerCase() else s

    requestHeaders.keys.map(k => (requestHeaders.get(k), pairs.get(k))).foldLeft(true) { (acc, pair) =>
      acc && (pair match {
        case (_, None) => {
          ignoreUnknown
        }
        case (v1, v2) => v1.map(maybeToLower).zip(v2.map(maybeToLower)).exists { p =>
          logger.info(s"Comparing request header [v1] with [v2].")

          p._1.matches(p._2)
        }
      })
    }
  }
}
